tao_compiler/mlir/disc/transforms/disc_comp_intens_fusion_to_cuda_source.cc

```diff
@@ -0,0 +1,631 @@
+  gather_b = isGatherB(func) ? "true" : "false";
+  scatter_d = isScatterD(func) ? "true" : "false";
+
+  std::string intent = "    ";
+  SourceEmitterCUDA source_emitter;
+  SourceEmitterCUDA::ValueNameBinding binding;
+  for (auto& op : func.getRegion().getOps()) {
+    if (isa<lmhlo::DotGeneralOp>(&op)) {
+      // Convert CUTLASS datatype to default cuda datatype for computation.
+      Value dot_output = op.getOperand(2);
+      std::string old_name = "input";
+      std::string new_name;
+      SmallVector<std::string> convert_instructions;
+      if (!mayConvertCutlassTypeToCUDAType(dot_output, old_name, new_name,
+                                           convert_instructions)) {
+        return false;
+      }
+      source_emitter.bindValueNames(SmallVector<Value>({dot_output}),
+                                    SmallVector<std::string>({new_name}),
+                                    binding);
+      for (auto& instruction : convert_instructions) {
+        specialized_epilogue += intent + instruction + ";\n";
+      }
+    } else if (!isa<func::ReturnOp>(&op)) {
+      assert(source_emitter.isSupportedOp(&op) && "Encounter unsupported op.");
+      auto instruction = source_emitter.EmitOp(&op, binding);
+      if (!instruction.hasValue()) {
+        return false;
+      } else {
+        specialized_epilogue += intent + instruction.value() + ";\n";
+      }
+    }
+  }
+  // Append return instruction.
+  SmallVector<Value> results;
+  getResults(func, results);
+  // Only support one result currently.
+  // TODO: support more outputs.
+  if (results.size() != 1) {
+    return false;
+  }
+  Value result = results[0];
+  std::string old_name = binding[result];
+  std::string new_name;
+  SmallVector<std::string> convert_result_instructions;
+  if (!mayConvertCUDATypeToCutlassType(result, old_name, new_name,
+                                       convert_result_instructions)) {
+    return false;
+  }
+  for (auto& instruction : convert_result_instructions) {
+    specialized_epilogue += intent + instruction + ";\n";
+  }
+  specialized_epilogue += intent + "return " + new_name + ";";
+
+  // Create source code op containing the source code string.
+  SmallVector<Value> operands;
+  getEffectiveOperands(func, operands);
+
+  int param_permute[3];
+  for (auto operand : llvm::enumerate(operands)) {
+    if (A == operand.value()) {
+      param_permute[0] = operand.index();
+    } else if (B == operand.value()) {
+      param_permute[1] = operand.index();
+    }
+  }
+  // TODO: identify the permutation if there are multiple results.
+  param_permute[2] = 2;
+  std::string parameter_permute = "{" + std::to_string(param_permute[0]) +
+                                  ", " + std::to_string(param_permute[1]) +
+                                  ", " + std::to_string(param_permute[2]) + "}";
+
+  // Replace newly generated code in the template.
+  stringReplaceInplace(cuda_code, kSpecializedClassName, specialized_class_name,
```

Comment:
A little advice for template-engine-like work is to put key-value pairs into a map, and render them with single function call to keep code shot and focus.

Reply:
would `absl::StrReplace` help?
https://github.com/abseil/abseil-cpp/blob/273292d1cfc0a94a65082ee350509af1d113344d/absl/strings/str_replace.h#L32-L37